### ISSUE TITLE
Trying to understand DenseDataLoader in `enzymes_diff_pool.py` example

### ISSUE THREAD

**Issue by tjiagoM** [rusty1s/pytorch_geometric#881]
## ❓ Questions & Help

Hi! Once more, thanks for this amazing tool!

I'm trying to understand the example in `examples/enzymes_diff_pool.py` example and I have a few questions which I'd really appreciate if you can answer me:

1. `DenseDataLoader` returns a `mask` for each batch, which (I think) happens because as you set the maximum number of nodes for each graph (100 in the example), and as each enzyme graph will have a different number of nodes, we need to know which ones actually correspond to the graph. However, the nodes which do not belong to a specific graph will just have a zero connection in the adjacency matrix anyway, right? Does this mask exist only for efficiency reasons for `DenseSAGEConv` and `dense_diff_pool` or some other reason? In other words, how bad would it be if I just ignore mask and set it to None?

2. After `dense_diff_pool` is used in line 105, mask is never used again. I guess that's because `adj` will have no connections for "invalid" nodes, but isn't this the same case as in my previous question? In other words, shouldn't `dense_diff_pool` also return a mask variable?

3. I was trying to use the `dense_diff_pool` with other methods from pytorch geometric so it would be very easy for me to use the normal (sparse) DataLoader, and just call `to_dense_adj` for each batch to get the adjacency matrices in correct form for diff pool. However, I've seen you saying in some other github issue that this is incorrect as the memory will explode. In your opinion what should be better: to use a sparse dataloader and call `to_dense_adj` when I used dense_diff_pool, or to use a dense dataloader and call `dense_to_sparse` when I use the other pytorch geometric operations (I've noticed you have a DenseGCNConv, but was wondering for others)? Or it is really one of those cases in which I just have to try?

Thanks!

**Comment by rusty1s** [comment]
Great questions!

1. The `mask` has shape `[batch_size, max_num_nodes]` and holds the information of a specific node originally exists in the sparse graph. It is in fact crucial to both dense message passing (node features of invalid nodes should stay zero) and diff pool (we do not want to cluster any of those invalid nodes). However, it should still work fine when ignoring the mask (node features and adjacency information are zero, so we are only injecting a bias into these features).

2. That is correct since `dense_diff_pool` creates exactly `num_clusters` nodes for each graph of which no node is invalid. So there is no need to make use of the mask again.

3. This is possible. You can use regular DataLoaders and sparse message passing operators before applying the first iteration of DiffPool. You need to make use of `to_dense_batch` and `to_dense_adj` though before calling DiffPool. After DiffPool, there is no (elegant) way to recover sparsity.

**Comment by tjiagoM** [comment]
Hey, thanks a lot for your prompt answers! That's what I wanted to know and now I can explore it further by myself, cheers!
